### ISSUE TITLE
basic bounds checking

### ISSUE THREAD

**Issue by lewis6991** [EmmyLuaLs/emmylua-analyzer-rust#524]
![Image](https://github.com/user-attachments/assets/a8755ae9-37a5-470e-9bc7-c3bc19ec5e6f)

```lua
local a --- @type string[]

if #a == 1 then
  --- @type string
  local _b = a[1]
end

if a[1] then
  --- @type string
  local _b = a[1]
end
```

If `a[1]` or `#a == 1` is true then `a[i]` cannot be `nil`.

**Comment by clason** [comment]
Similar issue:
```lua
local tasks ---@type function()
local  max_jobs = math.min(100, #tasks)
for i = 1, max_jobs do
      tasks[i]()
end
```
should not warn that "tasks[i] may be `nil`".

**Comment by xuhuanzy** [comment]
This requires the stream analysis refactoring to be completed. cppcxy is currently undergoing refactoring.

**Comment by lewis6991** [comment]
@clason I think that example is significantly more complicated and approaches the realms of formal analysis.

Realistically I think this might be the most we can hope for:

```lua
local tasks ---@type function()
local  max_jobs = math.min(100, #tasks)
for i = 1, max_jobs do
  if tasks[i] then
      tasks[i]()
  end
end
```

**Comment by clason** [comment]
Agree, but something like that should work (now that the flow refactor is merged):
```lua
for i = 1, #_G.arg do
  if _G.arg[i]:find('^--foo') then
    ...
  end
end
```

**Comment by lewis6991** [comment]
Yeah, that should work. It _might_ require an intermediate variable as this kind of analysis is usually limited to variables scoped locally, otherwise the complexity can increase too much.

```lua
local g_args = _G.arg
for i = 1, #g_args do
  if g_args[i]:find('^--foo') then
    ...
  end
end
```

**Comment by clason** [comment]
That's an acceptable workaround, but unfortunately doesn't work (yet) either.

**Comment by CppCXY** [comment]
> Yeah, that should work. It _might_ require an intermediate variable as this kind of analysis is usually limited to variables scoped locally, otherwise the complexity can increase too much.
> 
> local g_args = _G.arg
> for i = 1, #g_args do
>   if g_args[i]:find('^--foo') then
>     ...
>   end
> end

It doesn't have to be limited to local variables; in fact, simple indexing can also be checked.
